madminer/utils/morphing.py

```diff
@@ -123,41 +129,117 @@ def find_components(self, max_overall_power=4):
             The maximal sum of powers of all parameters contributing to the squared matrix element.
             Typically, if parameters can affect the couplings at n vertices, this number is 2n. Default value: 4.
 
+        BSM_max_power : int, optional
+            The maximal sum of powers of all parameters contributing to the squared matrix element for the BSM couplings.
+
+        Nd : int, optional
+            The number of decay vertices.
+
+        Np : int, optional
+            The number of production vertices.
+
+        Ns : int, optional
+            The number of couplings that works both for production and decay.
+
         Returns
         -------
         components : ndarray
             Array with shape (n_components, n_parameters), where each entry gives the power with which a parameter
             scales a given component.
         """
+        if Nd != 0 or Np != 0 or Ns != 0:
+            lst = []
+            # Check if any number of couplings were specified. 
+            if Nd == 0 and Np == 0 and Ns == 0:
+                raise RuntimeError("Coupling numbers not specified")
+
+            #number of couplings
+            gp = symbols('gp:15')
+            gd = symbols('gd:15')
+            gs = symbols('gs:15')
+
+            prod = sum(gp[:Np] + gs[:Ns])  #sum of couplings in production
+            dec = sum(gd[:Nd] + gs[:Ns])   #sum of couplings in decay
+
+            if (Nd==0 and Ns==0):
+                dec = 1
+            if (Np==0 and Ns==0):
+                prod = 1
+            f = expand((prod)**2*(dec)**2)  #contribution to matrix element squared
+            mono=Poly(f).terms(); #list of tuples containing monomials
+
+            for i in range(0, len(mono)):
+                lst.append(mono[i][0])
+
+            #array of coupligs powers in the alphabetic order gd0, gd1, ..., gp0, gp1, ..., gs0, gs1, ...
+            arr = np.array(lst)
+            len_arr = len(arr)
+
+
+            #cut over power_max
+            power_max = BSM_max_power
+
+            lst_pos = []
+
+            # Find the positions of the subarray that has elements exceed power_max
+            for j in range(0, len_arr):
+                for k in range(1, Nd):
+                    if(arr[j, k] > power_max):
+                        lst_pos.append(j)
+                    break    
+
+                for k in range(Nd+1, Nd+Np):
+                    if(arr[j, k] > power_max):
+                        lst_pos.append(j)
+                    break
```

Comment:
If this loop is iterating until it finds _\<something\>_, shouldn't it have the `break` clause inside the if?